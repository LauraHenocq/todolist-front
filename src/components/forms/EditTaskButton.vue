<template>
  <v-container>
    <v-btn data-cy="edit-task-button" color="pink">Modifier la t√¢che</v-btn>
  </v-container>
</template>