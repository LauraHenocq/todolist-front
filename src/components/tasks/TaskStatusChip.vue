<script setup lang="ts">
import { TaskStatus } from '@/models/task.entity';
import { computed } from 'vue';

type Props = {
  status: TaskStatus
}

const props = defineProps<Props>()

const getStatus = (): string => {
  switch (props.status) {
    case TaskStatus.TODO:
      return 'A faire'
    case TaskStatus.IN_PROGRESS:
      return 'En cours'
    case TaskStatus.DONE:
      return 'TerminÃ©e'
    default:
      return 'A faire'
  } 
}

const getStatusColor = computed(() => {
  switch (props.status) {
  case TaskStatus.TODO:
      return 'red'
    case TaskStatus.IN_PROGRESS:
      return 'secondary'
    case TaskStatus.DONE:
      return 'green'
    default:
      return 'grey'
  }
});
</script>

<template>
  <v-chip variant="flat" :color="getStatusColor">{{ getStatus(props.status) }}</v-chip>
</template>