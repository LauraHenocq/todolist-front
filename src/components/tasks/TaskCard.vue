<script setup lang="ts">
import { TaskEntity } from '@/models/task.entity';
import TaskStatusChip from '@/components/tasks/TaskStatusChip.vue';


type Props = {
  task: TaskEntity
}

const { task } = defineProps<Props>()

const emit = defineEmits(['delete-task', 'edit-task', 'go-to-task']);

const editTask = () => {
  emit('edit-task', task.id);
};

const deleteTask = () => {
  emit('delete-task', task.id);
};

const goToTask = () => {
  emit('go-to-task', task.id);
}

</script>
<template>
  <v-card>
    <div data-cy="task-card" class="task-card" @click="goToTask">
      <v-card-title>
        <span class="headline">{{ task.title }}</span>
      </v-card-title>
      <v-card-subtitle>
        <span>{{ task.description }}</span>
      </v-card-subtitle>
      <v-card-text>
        <TaskStatusChip :status="task.status" />
      </v-card-text>
    </div>
    <v-card-actions>
      <v-btn data-cy="task-card-edit-task-button" text @click="editTask(task)">Modifier</v-btn>
      <v-btn data-cy="task-card-delete-task-button" text @click="deleteTask(task.id)">Supprimer</v-btn>
    </v-card-actions>
  </v-card>
</template>
<style scoped>
.task-card {
  cursor: pointer;
}
</style>