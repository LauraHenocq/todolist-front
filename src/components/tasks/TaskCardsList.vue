<script setup lang="ts">
import { TaskEntity, TaskStatus } from '@/models/task.entity';
import TaskCard from '@/components/tasks/TaskCard.vue';

type Props = {
  tasks: TaskEntity[]
}

defineProps<Props>()

const emit = defineEmits(['task-clicked', 'delete-task']);

const emitTaskClicked = (taskId: string) => {
  emit('task-clicked', taskId)
};

const deleteTask = (taskId: string) => {
  emit('delete-task', taskId);
}

</script>

<template>
<v-container>
    <v-row>
      <v-col v-for="task in tasks" :key="task.id" cols="12" md="6">
        <TaskCard :task="task" @deleteTask="deleteTask"/>
      </v-col>
    </v-row>
  </v-container>
</template>
