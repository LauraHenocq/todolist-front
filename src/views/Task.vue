<script setup lang="ts">
import { onMounted, ref } from 'vue';
import { useRoute } from 'vue-router';
import { useTasksStore } from '@/stores/tasks.store';
import { storeToRefs } from 'pinia';
import TaskStatusChip from '@/components/tasks/TaskStatusChip.vue';
import EditTaskButton from '@/components/forms/EditTaskButton.vue';
import type { UpdateTaskInput } from '../services/inputs/update-task.input';
import TaskEditionDialog from '@/components/dialogs/TaskEditionDialog.vue';
import Alert from '@/components/common/Alert.vue';
import { TaskEntity } from '@/models/task.entity';

const tasksStore = useTasksStore();
const { task } = storeToRefs(tasksStore);
const { handleUpdateTask, getTaskById } = tasksStore;

const isLoading = ref(false);

const route = useRoute();
const taskId = ref(route.params.id);
const taskLocale = ref<TaskEntity | null>(null);

const alertInfo = ref({
  title: '',
  text: '',
  type: 'info'
});

const isAlertVisible = ref(false);

// TASK EDITION
const isTaskEditionDialogOpen = ref(false);

const openTaskEditionDialog = () => {
  isTaskEditionDialogOpen.value = true;
}
const closeTaskEditionDialog = () => {
  isTaskEditionDialogOpen.value = false;
}

const updateTask = async (updateTaskInput: UpdateTaskInput) => {
  isLoading.value = true;
  try {
    await handleUpdateTask(task.value.id, updateTaskInput);
    alertInfo.value = {
      title: 'Tâche mise à jour',
      text: 'La tâche a été mise à jour avec succès.',
      type: 'success'
    };
    isAlertVisible.value = true;
  } catch {
    alertInfo.value = {
      title: 'Erreur lors de la mise à jour',
      text: 'Une erreur est survenue lors de la mise à jour de la tâche, réessaie.',
      type: 'error'
    };
    isLoading.value = false;
    isAlertVisible.value = true;
  }
  
  closeTaskEditionDialog();
}

onMounted(async () => {
  isLoading.value = true;
  await getTaskById(taskId.value);
  isLoading.value = false;
});

</script>

<template>
  <v-container>
    <Alert v-if="isAlertVisible" :type="alertInfo.type" :title="alertInfo.title" :text="alertInfo.text" @close="isAlertVisible = false" />
    <div v-if="isLoading">
      <v-progress-circular
        color="primary"
        indeterminate
      ></v-progress-circular>
    </div>
    <div v-else-if="task">
      <TaskEditionDialog :is-open="isTaskEditionDialogOpen" :task=task @updateTask="updateTask" @cancel="closeTaskEditionDialog" />
      <p class="my-lg-8 my-4 text-lg-h4 text-h6 font-weight-medium">Détail d'une tâche</p>
      <p class="my-lg-8 my-4 text-lg-h3 text-h5 font-weight-medium">{{ task.title }}</p>
      <TaskStatusChip :status="task.status" />
      <p class="my-lg-8 my-4">{{ task.description }}</p>
      <EditTaskButton class="my-lg-8 my-4 text-center" @click="openTaskEditionDialog" />
    </div>
  </v-container>
</template>