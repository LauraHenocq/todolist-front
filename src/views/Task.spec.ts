import { describe, it, expect, vi, beforeEach } from 'vitest';
import Task from '@/views/Task.vue';
import { createTestingPinia } from '@pinia/testing';
import { nextTick } from 'vue';
import { useTasksStore } from '@/stores/tasks.store';
import { TaskEntity, TaskStatus } from '@/models/task.entity';
import { mount, VueWrapper } from '@vue/test-utils';
import { createVuetify } from 'vuetify';
import EditTaskButton from '@/components/forms/EditTaskButton.vue';

const vuetify = createVuetify();

const mockUseRouter = {
  push: vi.fn(),
};

const mockUseRoute = {
    params: { id: 'task-id' },
  };
  
vi.mock('vue-router', () => ({
  useRoute: () => mockUseRoute,
  useRouter: () => mockUseRouter,
}));

describe('TaskView', () => {
  let wrapper: VueWrapper;
  const pinia = createTestingPinia();
  let tasksStore: ReturnType<typeof useTasksStore>;
  
  global.ResizeObserver = require('resize-observer-polyfill');
  
  beforeEach(() => {
    wrapper = mount(Task, {
      global: {
        plugins: [pinia, vuetify],
      },
    });
    tasksStore = useTasksStore();
    tasksStore.task = new TaskEntity({ title: 'Task Title', description: 'Task Description', _status: TaskStatus.TODO, id: 'task-id' });
  });

  describe('render', () => {
    it('should render', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
  
    it('should show loading spinner when data is loading', async () => {
      wrapper.vm.isLoading = true;
      await nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('Task retrieving', () => {
    it('should call getTaskById with the id params on mounted', async () => {
      expect(tasksStore.getTaskById).toHaveBeenCalledWith('task-id');
    });
  });

  describe('edit the task', () => {
    it('should open the edit task dialog form when the edit task button is clicked', async () => {
      await wrapper.findComponent(EditTaskButton).trigger('click');
      expect(wrapper.vm.isTaskEditionDialogOpen).toBe(true);
    });

    it('should edit the task and show success alert when the user validates the edition', async () => {
      const updateTaskInput = { title: 'Titre de la nouvelle tâche', description: 'Description de la nouvelle tâche', status: TaskStatus.DONE }
      
      await wrapper.vm.openTaskEditionDialog();
            
      await wrapper.vm.updateTask(updateTaskInput);
      
      expect(tasksStore.handleUpdateTask).toHaveBeenCalled();
      expect(tasksStore.handleUpdateTask).toHaveBeenCalledWith('task-id', updateTaskInput);
      expect(wrapper.vm.alertInfo.title).toBe('Tâche mise à jour');
      expect(wrapper.vm.isAlertVisible).toBe(true);
      });
  });
});