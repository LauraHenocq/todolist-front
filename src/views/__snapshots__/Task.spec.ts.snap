// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskView > render > should render 1`] = `
"<div class="v-container v-locale--is-ltr">
  <!--v-if-->
  <div>
    <!---->
    <!---->
    <p class="my-lg-8 my-4 text-lg-h4 text-h6 font-weight-medium">Détail d'une tâche</p>
    <p class="my-lg-8 my-4 text-lg-h3 text-h5 font-weight-medium">Task Title</p><span class="v-chip v-theme--light bg-red v-chip--density-default v-chip--size-default v-chip--variant-flat" draggable="false"><!----><span class="v-chip__underlay"></span>
    <!---->
    <!---->
    <div class="v-chip__content" data-no-activator="">A faire</div>
    <!---->
    <!----></span>
    <p class="my-lg-8 my-4">Task Description</p>
    <div class="v-container v-locale--is-ltr my-lg-8 my-4 text-center"><button type="button" class="v-btn v-btn--elevated v-theme--light bg-pink v-btn--density-default v-btn--size-default v-btn--variant-elevated" data-cy="edit-task-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
        <!----><span class="v-btn__content" data-no-activator="">Modifier la tâche</span>
        <!---->
        <!---->
      </button></div>
  </div>
</div>"
`;

exports[`TaskView > render > should show loading spinner when data is loading 1`] = `
"<div class="v-container v-locale--is-ltr">
  <!--v-if-->
  <div>
    <div class="v-progress-circular v-progress-circular--indeterminate v-theme--light v-progress-circular--size-default text-primary" role="progressbar" aria-valuemin="0" aria-valuemax="100"><svg style="transform: rotate(calc(-90deg + 0deg));" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 45.714285714285715 45.714285714285715">
        <circle class="v-progress-circular__underlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.66370614359172" stroke-dashoffset="0"></circle>
        <circle class="v-progress-circular__overlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.66370614359172" stroke-dashoffset="125.66370614359172px"></circle>
      </svg>
      <!---->
    </div>
  </div>
</div>"
`;
